File: dataset/java/C05PalindromeType4.java
import java.util.Deque;
import java.util.LinkedList;

public class C05PalindromeType4 {
    public static boolean isPalindrome(String text) {
        Deque<Character> dq = new LinkedList<>();
        for (char ch : text.toCharArray()) {
            if (Character.isLetterOrDigit(ch)) {
                dq.addLast(Character.toLowerCase(ch));
            }
        }

        while (dq.size() > 1) {
            if (!dq.removeFirst().equals(dq.removeLast())) {
                return false;
            }
        }

        return true;
    }
}

File: test/javascript/mix/C05PalindromeType4.js
function isPalindrome(text) {
    const dq = [];

    for (const ch of text) {
        if (/[a-zA-Z0-9]/.test(ch)) {
            dq.push(ch.toLowerCase());
        }
    }

    while (dq.length > 1) {
        if (dq.shift() !== dq.pop()) {
            return false;
        }
    }

    return true;
}

module.exports = { isPalindrome };
